modules/core/src/main/scala/dk/cwconsult/peregrin/core/impl/dao/ChangeLogDAOImpl.scala

```diff
@@ -0,0 +1,144 @@
+package dk.cwconsult.peregrin.core.impl.dao
+
+import java.sql.Connection
+import java.util.UUID
+
+import dk.cwconsult.peregrin.core.Schema
+import dk.cwconsult.peregrin.core.Table
+import dk.cwconsult.peregrin.core.impl.ChangeLogEntry
+import dk.cwconsult.peregrin.core.impl.dao.BaseDAO.ColumnProbe.Missing
+
+import scala.collection.mutable.ArrayBuffer
+
+class ChangeLogDAOImpl(schema: Schema, connection: Connection)
+  extends BaseDAO(connection)
+    with ChangeLogDAO {
+
+  import dk.cwconsult.peregrin.core.impl.ConnectionImplicits._
+
+  private[this] val changeLogTable =
+    Table(
+      name = "__peregrin_changelog__",
+      schema)
+
+  private[this] val createPeregrinSchema: String =
+    s"CREATE SCHEMA IF NOT EXISTS $schema"
+
+  private[this] val createPeregrinChangeLogTable: String =
+    s"""
+        CREATE TABLE IF NOT EXISTS $changeLogTable (
+          "identifier" INT NOT NULL,
+          "sql" TEXT NOT NULL,
+          "executed" TIMESTAMP WITH TIME ZONE NOT NULL)
+      """.stripMargin
+
+  private[this] val alterPeregrinChangeLogTableAddUUIDId: String =
+    s"""
+        ALTER TABLE $changeLogTable
+        ADD COLUMN "migration_id" UUID
+     """.stripMargin
+
+  /**
+    * Create change log if necessary.
+    */
+  override def createChangeLogIfMissing(): Unit = {
+    // Create the schema if necessary.
+    connection.execute(createPeregrinSchema)
+    // Create the migration log table.
+    val _ = connection.execute(createPeregrinChangeLogTable)
+    // Migrate the schema, based on which columns we have available
+    probeAndMigrateTable(changeLogTable)(
+      // Add new UUID identifier for migration, if it doesn't exist already
+      Missing("migration_id") -> {
+        connection.execute(alterPeregrinChangeLogTableAddUUIDId)
+      })
```

Comment:
Jeg tror ikke jeg forstår motivationen for at gøre det med probeAndMigrate.

Kan du ikke bare lave en funktion der hedder "getChangelogVersion" der checker på om "version" findes i changelog_version tabellen, og antager at det versionsnummer er korrekt? Hvis tabellen eller rækken ikke findes, så antag version 0 (den før dine ændringer).

Derefer: Kør slavisk den SQL der skal til for at komme fra version X til Y (seneste). Disse meta-migrationer på selve changeloggen behøver slet ikke blive gemt eller have noget persistent id. (For de er 100% under kontrol af peregrin selv, så vi *ved* at de ikke ændrer sig for vi styrer peregrin.)

Det jeg er bekymret for her at logikken omkring at detektere hvad der skal laves bliver arbitrært kompleks og den falder faktisk i den samme problematik som migrering basere på rene "deltaer" mellem to skemaer.

Reply:
Jeg tror måske der er noget misforståelse her, jeg forstår i hvert fald ikke lige relevansen af version.

Det koden her gør, er at probe tabellen for, om kolonnen migration_id findes, eller om den skal tilføjes med en ALTER TABLE. Jeg har pg9.5 som target, så jeg kan vist ikke bruge IF NOT EXISTS.

Jeg kan selvfølgelig godt probe tabellen på andre måder end jeg gør, men andre metoder jeg forsøgte fejlede gerne hårdt, så det var problematisk at fortsætte.

Reply:
OK, bare for at vi har samme opfattelse, så opsummerer jeg lige hvad jeg lidt antog var modellen for håndteringen af migrering af selve changelog-formatet:

- Vi opretter en ny tabel peregrin_changelog_metadata (IF NOT EXISTS) der indeholder key-value par. I den putter vi "version -> ...". [Jeg er forholdvis ligeglad med hvad den indeholder -- pointen er at vi kan se format-version deri.]
- Ved opstart kigger Peregrin *kun* på versionsnummer og migrerer fra det der står til "seneste" version ved bare at apply'e en liste migreringer defineret ud fra versionsnumre. Så de her migreringer vil være noget i stil med

     ``` 
         changeLogMigrations = Seq(
            1 -> Seq(
               "ALTER TABLE ...",
               "ADD COLUMN ...",
            ),
            2 -> Seq(
               "..."
            )
         )
   ```

(Hvis der ikke er nogen peregrin_changelog_metadata tabel, så antages version 0.)

Giver det mening?

Det der er problematisk ved probe...() metoden er at man bliver nødt til at lave heuristikker for at detektere hvad der mangler at blive lavet. (Svjks er der måske også edge cases hvis man f.eks. beslutter sig for at fjerne missing_id kolonnen, og så senere igen gerne vil tilføje. Hvis klienten laver migrering hele vejen så vil data i kolonnen blive bevaret, men det *burde* have været slettet.)

 Med en "format-version" på changeloggen er det *altid* 100% forudsigeligt hvad der vil ske og logikken er universel og simpel.

Reply:
Det giver mere mening, hvad du tænker. Men det er ikke det _det_ versionsnummer er til for, det vil i så fald være et andet versionsnummer.
Versionsnummeret angiver reelt hvilken Migration-type user applikationen/library'et bruger, og dermed hvilken information der skal bruges fra changelog tabellen. Du snakker om et versionsnummer for formatversionen, det vil i så fald være separat.
Et eksempel: Mit næste PR er jo namespaces, og der vil det også være nødvendigt at tilføje en kolonne, men den changelog version der bruges på baggrund af brugerens migrations, ændrer sig ikke.
Vi kan godt have et separat versionsnummer, men så vidt jeg ser bliver denne logik bare mere kompliceret af det, da du i stedet for at skulle agere på baggrund af det du mangler, nu skal til at implementere handlinger for version 1, 2, 3..
Det bliver heller ikke simplere af, at der kan være flere versioner af lib'et, som tilgår samme db tabel, så det kan være ukendte versionsnumre. Derfor synes jeg det er simplere, at hver migration ser om den har det den skal bruge, og ellers opretter det.

Reply:
Jeg er helt med på at MigrationVn er nogen 'separate' versionsnumre i den forstand at vi skal kunne håndtere V1, V2, ... men det skal vi *alligevel*.

> Vi kan godt have et separat versionsnummer, men så vidt jeg ser bliver denne logik bare mere kompliceret af det, da du i stedet for at skulle agere på baggrund af det du mangler, nu skal til at implementere handlinger for version 1, 2, 3..

Nej, kun handlinger for at komme fra changelog-fmt 1 til changelog-fmt 2, fra changelog-fmt 2 til 3, etc.

(Jeg snakker iøvrigt *kun* om selve changelog-formatet her, bare sådan for at gøre det helt klart.)

> Det bliver heller ikke simplere af, at der kan være flere versioner af lib'et, som tilgår samme db tabel, så det kan være ukendte versionsnumre. Derfor synes jeg det er simplere, at hver migration ser om den har det den skal bruge, og ellers opretter det.

Hvis den ser et ukendt versionsnummer så skal den baile ud *med det samme*. Der er ingen sane måde at håndtere den situation uanset hvad man gør.

Reply:
Det var måske ikke helt klart med eksempelkoden, men idéen er at det nummer der står til venstre for "->" er versionsnummeret som man er på efter at have kørt den pågældende SQL *og alle foregående* SQL statements for versionerne før det. (Så det fungerer præcis som peregrin gør nu, bare uden persistens af migrations.)